dbt/include/spark/macros/materializations/incremental.sql

```diff
@@ -100,9 +101,11 @@
     {% do dbt_spark_validate_merge(file_format) %}
   {% endif %}
 
-  {% call statement() %}
-    set spark.sql.sources.partitionOverwriteMode = DYNAMIC
-  {% endcall %}
+  {% if unique_key or config.get('partition_by') %}
```

Comment:
I think we're almost there.
```suggestion
  {% if config.get('partition_by') %}
```
I don't think we need to set this if we use an `unique_key`. This one is only related to the `MERGE INTO` strategy, and not related to overwriting partitions.

Reply:
But doesnt the merge strategy also need the dynamic overwrite setting?

Reply:
It only affects if you have a partitioned table. Some background information:

If you do:

```sql
SET spark.sql.sources.partitionOverwriteMode=static;
```

Requires you to set the partition explicit:

```sql
INSERT OVERWRITE TABLE stats
PARTITION(country = 'US', year = 2017, month = 3, day)
SELECT ad, SUM(impressions), SUM(clicks), log_day
FROM impression_logs
GROUP BY ad;
```

Setting it to Dynamic will enable Spark to get the data from the query itself:

```sql
SET spark.sql.sources.partitionOverwriteMode=dynamic;
```

This query will check which values there are in the partitioned columns, and then replace the existing data with new data:

```
INSERT OVERWRITE TABLE stats
SELECT ad, SUM(impressions), SUM(clicks), log_day
FROM impression_logs
WHERE country = 'US' AND year = 2017 AND month = 3
GROUP BY ad;
```

These queries will have the same effect, but it is easy to lose data if you accidentally replace a partition that you didn't intend to. Therefore you have to enable the `dynamic` behavior explicitly.

With the merge-into statement:
```
Copy to clipboardCopy
MERGE INTO [db_name.]target_table [AS target_alias]
USING [db_name.]source_table [AS source_alias]
ON <merge_condition>
[ WHEN MATCHED [ AND <condition> ] THEN <matched_action> ]
[ WHEN MATCHED [ AND <condition> ] THEN <matched_action> ]
[ WHEN NOT MATCHED [ AND <condition> ]  THEN <not_matched_action> ]
```

It will insert or replace using the `unique_key` in the upsert the data in the table. This is unrelated to the partitions. So, if you have the table partitioned on a date, it will potentially upsert records into multiple partitions.